### PR TITLE
Initial commit of Interpretability notebook

### PR DESCRIPTION
**Pull request by cmurray-astro**
This is the initial commit of the Interpretability notebook by Oliver Lin and Daisuke Nagai. @oliverlin23

---------
This notebook checklist has been made available to us by the [Notebooks For All team](https://github.com/Iota-School/notebooks-for-all/blob/main/resources/event-hackathon/notebook-authoring-checklist.md).
Its purpose is to serve as a guide for both the notebook author and the technical reviewer highlighting critical aspects to consider when striving to develop an accessible and effective notebook.

### The First Cell

- [x] The title of the notebook in a first-level heading (eg. `<h1>` or `# in markdown`).
- [x] A brief description of the notebook.
- [x] A table of contents in an [ordered list](https://www.markdownguide.org/basic-syntax/#ordered-lists) (`1., 2.,` etc. in Markdown).
- [x] The author(s) and affiliation(s) (if relevant).
- [x] The date first published.
- [x] The date last edited (if relevant).
- [x] A link to the notebook's source(s) (if relevant).

### The Rest of the Cells

- [x] There is only one H1 (`#` in Markdown) used in the notebook.
- [x] The notebook uses other [heading tags](https://www.markdownguide.org/basic-syntax/#headings) in order (meaning it does not skip numbers). 

## Text

- [x] All link text is descriptive. It tells users where they will be taken if they open the link.
- [x] All acronyms are defined at least the first time they are used. 
- [x] Field-specific/specialized terms are used when needed, but not excessively.

## Code

- [x] Code sections are introduced and explained before they appear in the notebook. This can be fulfilled with a heading in a prior Markdown cell, a sentence preceding it, or a code comment in the code section.
- [x] Code has explanatory comments (if relevant). This is most important for long sections of code.
- [x] If the author has control over the syntax highlighting theme in the notebook, that theme has enough color contrast to be legible.
- [x] Code and code explanations focus on one task at a time. Unless comparison is the point of the notebook, only one method for completing the task is described at a time.

### Images

- [ ] All images (jpg, png, svgs) have an [image description](https://www.w3.org/WAI/tutorials/images/decision-tree/). This could be
    - [ ] [Alt text](https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/alt) (an `alt` property) 
    - [ ] [Empty alt text](https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/alt#decorative_images) for decorative images/images meant to be skipped (an `alt` attribute with no value)
    - [ ] Captions
    - [ ] If no other options will work, the image is decribed in surrounding paragraphs.

- [ ] Any [text present in images](https://www.w3.org/WAI/WCAG21/Understanding/images-of-text.html) exists in a text form outside of the image (this can be alt text, captions, or surrounding text.)

### Visualizations

- [ ] All visualizations have an image description. Review the previous section, Images, for more information on how to add it.
- [ ] [Visualization descriptions](http://diagramcenter.org/specific-guidelines-e.html) include
    - [ ] The type of visualization (like bar chart, scatter plot, etc.)
    - [ ] Title
    - [ ] Axis labels and range
    - [ ] Key or legend
    - [ ] An explanation of the visualization's significance to the notebook (like the trend, an outlier in the data, what the author learned from it, etc.)

- [ ] All visualizations and their parts have [enough color contrast](https://www.w3.org/WAI/WCAG21/Understanding/contrast-minimum.html) ([color contrast checker](https://webaim.org/resources/contrastchecker/)) to be legible. Remember that transparent colors have lower contrast than their opaque versions.
- [ ] All visualizations [convey information with more visual cues than color coding](https://www.w3.org/WAI/WCAG21/Understanding/use-of-color.html). Use text labels, patterns, or icons alongside color to achieve this.
- [ ] All visualizations have an additional way for notebook readers to access the information. Linking to the original data, including a table of the data in the same notebook, or sonifying the plot are all options.

### CODE REVIEW

**Comment by cmurray-astro** [comment]
hi @oliverlin23 thanks for contributing this notebook! The first question I have is can you please provide a "requirements" file which specifies the versions of all required packages? I ran into an error with tensorflow when trying to run the notebook.

**Comment by cmurray-astro** [comment]
or, to make it simpler, can you tell me which versions of the required packages you have installed locally?

**Comment by oliverlin23** [comment]
Hi Claire, thanks for letting me know! My computer is running Keras 2.12.0, Tensorflow 2.12.1, Astropy 5.3.3, numpy 1.25.2, matplotlib 3.7.2, opencv 4.8.0, and scikit-learn 1.2.2. Could you let me know if it still has errors? Thank you!

**Comment by cmurray-astro** [comment]
thanks! I put together a requirements file based on those package versions and it all runs well now. My next step is to do a code and style review, and I will be back in touch with questions and comments! :)

File: notebooks/hello-universe/Interpreting_CNNs/Interpreting_CNNs.ipynb [hunk 1]
```diff
@@ -0,0 +1,803 @@
...
+   "source": [
+    "**Image caption**: The above image contains three panels in a horizontal row. The first panel shows the original image of a merginig galaxy candidate, the second panel shows a logarithmically-scaled version of the original image, and the third panel shows a heat map of the grad-cam map."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "697a59ff-53b3-4353-804c-5061281d5b3a",
+   "metadata": {},
+   "source": [
+    "**The result is very similar** to our saliency map from backpropagation, with the image being coarser but also showing bot the top and bottom of the galaxy. We can also play around with the selected layer to calculate the output at different steps in the model and see how activation changes throughout the model."
```
Comment (eteq) [inline notebooks/hello-universe/Interpreting_CNNs/Interpreting_CNNs.ipynb:519]:
```suggestion
    "**The result is very similar** to our saliency map from backpropagation, although the image is coarser and also showing both the top and bottom of the galaxy. We can also play around with the selected layer to calculate the output at different steps in the model and see how activation changes throughout the model."
```

File: notebooks/hello-universe/Interpreting_CNNs/Interpreting_CNNs.ipynb [hunk 2]
```diff
@@ -0,0 +1,803 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "id": "27909148-c421-48b6-9ecb-a47f56a5dd61",
+   "metadata": {
+    "tags": []
+   },
+   "source": [
+    "## 3. RISE Algorithm\n",
+    "The RISE (Randomized Input Sampling for Explanation) Algorithm is another interpretation technique for calculating saliency maps. Instead of calculating gradients from within the model, the RISE implementation works by covering up pieces of the input image, running it through the model, and calculating the average activation in order to determine what parts of the image are most important. As such, this method does not require any access to the inner workings of the model. The algorithm first generates a sequence random of binary grids (called masks), which are placed onto the image. Everything not covered by the mask is removed by multiplying the images together, and the resultant activations are averaged to get our final heatmap. A full description of the algorithm and its variations is provided by <a href='https://arxiv.org/abs/1806.07421'>Petsiuk et al. 2018</a>. \n",
```
Comment (eteq) [inline notebooks/hello-universe/Interpreting_CNNs/Interpreting_CNNs.ipynb:530]:
```suggestion
    "The RISE (Randomized Input Sampling for Explanation) Algorithm is another interpretation technique for calculating saliency maps. Instead of calculating gradients from within the model, the RISE implementation works by covering up pieces of the input image, running it through the model, and calculating the average activation in order to determine what parts of the image are most important. As such, this method does not require any access to the inner workings of the model. The algorithm first generates a random sequence of binary grids (called masks), which are placed onto the image. Everything not covered by the mask is removed by multiplying the images together, and the resultant activations are averaged to get our final heatmap. A full description of the algorithm and its variations is provided by <a href='https://arxiv.org/abs/1806.07421'>Petsiuk et al. 2018</a>. \n",
```

File: notebooks/hello-universe/Interpreting_CNNs/Interpreting_CNNs.ipynb [hunk 3]
```diff
@@ -0,0 +1,803 @@
...
+    "    masks[i] = mask\n",
+    "\n",
+    "masks = masks.reshape(-1, *input_shape[:2], 1)\n",
+    "\n",
+    "N = len(masks)\n",
+    "pred_masks = cnn.predict(image * masks)\n",
+    "pred_masks = np.expand_dims(pred_masks, axis=-1)\n",
+    "pred_masks = np.expand_dims(pred_masks, axis=-1) # Reshape pred_masks for broadcasting\n",
+    "heatmap = (pred_masks * masks).sum(axis=0)\n",
+    "heatmap = heatmap / N / p1"
```
Comment (eteq) [inline notebooks/hello-universe/Interpreting_CNNs/Interpreting_CNNs.ipynb:572]: Is this baked into to Keras or tensorflow somewhere?  While the previous two are straightforward, this seems complex enough it could benefit from an implementation in a library.  It still makes sense to leave this here to help the reader understand, but if there's a way to make it using a library implementation perhaps that should be mentioned?

Comment (eteq) [reply notebooks/hello-universe/Interpreting_CNNs/Interpreting_CNNs.ipynb:572]: (To be clear I'm just guessing there's some way to do it from a library.  If not, just ignore this!)

File: notebooks/hello-universe/Interpreting_CNNs/Interpreting_CNNs.ipynb [hunk 4]
```diff
@@ -0,0 +1,803 @@
...
+   "execution_count": null,
+   "id": "a8a6e994-1360-45d8-a6a4-0f8e3bc9bb24",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Plot the results next to the original image\n",
+    "fig, axes = plt.subplots(1, 2, figsize=(14, 5))\n",
+    "axes[0].imshow(image)\n",
+    "axes[0].set_title(\"orig_img\")\n",
+    "i = axes[1].imshow(heatmap, cmap=\"jet\", alpha=0.8)\n",
```
Comment (eteq) [inline notebooks/hello-universe/Interpreting_CNNs/Interpreting_CNNs.ipynb:586]:
```suggestion
    "i = axes[1].imshow(heatmap, cmap=\"turbo\")\n",
```
or
```suggestion
    "i = axes[1].imshow(heatmap)\n",
```

File: notebooks/hello-universe/Interpreting_CNNs/Interpreting_CNNs.ipynb [hunk 5]
```diff
@@ -0,0 +1,803 @@
...
+   "source": [
+    "**Image caption**: The above image contains two panels in a horizontal row. The first panel shows the original image of a merginig galaxy candidate, and the second panel shows a heat map of the RISE map."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "93bcaa55-1884-4665-8040-61e16ab35c90",
+   "metadata": {},
+   "source": [
+    "**When examining the selected image**, we see a similar ring around the galactic center. Note that this is not the case for all astronomical images, or even all images in this dataset. Try playing around with the selected image to generate different saliency maps. For images where the central feature is relatively small, RISE's occlusion-based methodology can be unreliable."
```
Comment (eteq) [inline notebooks/hello-universe/Interpreting_CNNs/Interpreting_CNNs.ipynb:604]: "similar" to what?  I don't really understand what that's saying.

File: notebooks/hello-universe/Interpreting_CNNs/Interpreting_CNNs.ipynb [hunk 6]
```diff
@@ -0,0 +1,803 @@
...
+    "\n",
+    "In the exercise below, we will build and train four mini-models on the same data set as before. As this is an educational notebook, we will limit the training time of each model to five epochs. Results with these models may vary considerably due to these training constraints, but we highly encourage you to try modifying this section of the notebook for different results (see [Exercises](#Exercises/Extensions))\n",
+    "\n",
+    "Note that performing an ablation analysis will require training multiple models with the same architecture. This can be quite power intensive on personal computers, so it is recommended that devices be plugged in before running the analysis.\n",
```
Comment (eteq) [inline notebooks/hello-universe/Interpreting_CNNs/Interpreting_CNNs.ipynb:619]:
```suggestion
    "Note that performing an ablation analysis will require training multiple models with the same architecture. This can be quite compute intensive, so if you are running this notebook locally it is recommended that your device be plugged in before running the analysis.\n",
```
I get that the point here is "beware running on your laptop", but I think it's still more the point that it's compute intensive which means you might need your laptop plugged in.

File: notebooks/hello-universe/Interpreting_CNNs/Interpreting_CNNs.ipynb [hunk 7]
```diff
@@ -0,0 +1,803 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "3e0a2ddf-665a-4ddc-a69b-417f149253b6",
+   "metadata": {},
+   "source": [
+    "## FAQ\n",
+    "\n",
+    "- **How do I choose which saliency model to use?** Each saliency technique has it's own strengths and weaknesses. If you are looking for an intuitive approach that produces the highest resolution image, then gradient backpropagation is the best choice. If you want to control which layer to look at and watch how the model evolves or are concerned about reliability, then Grad-CAM will perform better. If you don't have access to the inner workings of the model or want to understand how a model would perform with slightly different inputs, then RISE is the best option.\n",
```
Comment (eteq) [inline notebooks/hello-universe/Interpreting_CNNs/Interpreting_CNNs.ipynb:716]: This doesn't mention when you'd want to use ablation analysis.  Is that intentional?  If I just read this knowing nothing but this notebook about these techniques (which is basically true for me...) I might conclude there's no use to ever doing ablation analysis.

File: notebooks/hello-universe/Interpreting_CNNs/Interpreting_CNNs.ipynb [hunk 8]
```diff
@@ -0,0 +1,803 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "476f94b4-b2d8-4c20-8db8-09d02d018938",
+   "metadata": {},
+   "outputs": [],
+   "source": []
+  }
```
Comment (eteq) [inline notebooks/hello-universe/Interpreting_CNNs/Interpreting_CNNs.ipynb:780]: I think this is an accidental blank cell at the end?
```suggestion
  }
```

**Review by eteq — CHANGES_REQUESTED** [review]
This is a great notebook overall, I learned a lot by reviewing it!

That said, I found some typos and several areas where I was having trouble understanding the wording.  I also have a few style items I think need addressing.

It's all inline, though, showing that the big picture of this notebook is great!

File: notebooks/hello-universe/Interpreting_CNNs/Interpreting_CNNs.ipynb (hunk 7 above)
Comment (cmurray-astro) [reply notebooks/hello-universe/Interpreting_CNNs/Interpreting_CNNs.ipynb:716]: @oliverlin23 please provide a citation to ablation analysis for interpretability, along with a short sentence describing when it could be useful

Comment (cmurray-astro) [reply notebooks/hello-universe/Interpreting_CNNs/Interpreting_CNNs.ipynb]: resolved via Slack conversation, associated edits incorporated.

**Comment by eteq** [comment]
This all looks great, thanks @cmurray-astro ! Ready to merge.

**Review by eteq — APPROVED** [review]
